File: app/src/main/java/com/jupiterreporting/app/MainActivity.java
package com.jupiterreporting.app;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.jupiterreporting.auth.AuthManager;
import com.jupiterreporting.data.database.DatabaseProvider;
import com.jupiterreporting.data.entities.Report;
import com.jupiterreporting.sync.SyncWorker;

import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {

    private AuthManager authManager;
    private GoogleSignInAccount account;
    private String spreadsheetId = "1J5wxqk1_nCPEUBnilSHI8RgnWU7CUV-vxrAVGl6LX5M";
    private String sheetName = "Sheet1";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Initialize AuthManager
        authManager = new AuthManager(this);
        authManager.signIn(this);

        // Find UI elements
        Button saveButton = findViewById(R.id.saveButton);
        EditText field1EditText = findViewById(R.id.field1EditText);
        EditText field2EditText = findViewById(R.id.field2EditText);

        // Set save button click listener
        saveButton.setOnClickListener(v -> {
            String field1 = field1EditText.getText().toString();
            String field2 = field2EditText.getText().toString();
            saveReport(field1, field2);
        });

        // Schedule WorkManager for periodic sync
        PeriodicWorkRequest syncWorkRequest =
                new PeriodicWorkRequest.Builder(SyncWorker.class, 15, TimeUnit.MINUTES)
                        .build();
        WorkManager.getInstance(this).enqueue(syncWorkRequest);
    }

    // Handle sign-in result
    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        authManager.handleSignInResult(requestCode, resultCode, data, new AuthManager.SignInCallback() {
            @Override
            public void onSuccess(GoogleSignInAccount account) {
                MainActivity.this.account = account;
                // Now you can use the account to access Google Sheets API
            }

            @Override
            public void onFailure(Exception e) {
                // Handle sign-in failure
            }
        });
    }

    // Save report to the local database
    private void saveReport(String field1, String field2) {
        Report report = new Report();
        report.setField1(field1);
        report.setField2(field2);
        report.setIsSynced(false);

        DatabaseProvider.getDatabase(this).reportDao().insert(report);
    }
}

File: app/src/main/java/com/example/jupiter_reporting/MainActivity.kt
package com.example.jupiter_reporting

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.auth.api.signin.*
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.Scope
import com.google.android.gms.tasks.Task
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.http.javanet.NetHttpTransport
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.sheets.v4.Sheets
import com.google.api.services.sheets.v4.SheetsScopes
import android.util.Log

class MainActivity : AppCompatActivity() {

    private lateinit var googleSignInClient: GoogleSignInClient
    private lateinit var credential: GoogleAccountCredential
    private lateinit var sheetsService: Sheets

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestEmail()
            .requestScopes(Scope(SheetsScopes.SPREADSHEETS))  // Используем объект Scope с областью доступа
            .build()

        googleSignInClient = GoogleSignIn.getClient(this, gso)

        val account = GoogleSignIn.getLastSignedInAccount(this)
        if (account != null) {
            initializeSheetsApi(account)
        } else {
            signIn()
        }
    }

    private fun signIn() {
        val signInIntent = googleSignInClient.signInIntent
        startActivityForResult(signInIntent, RC_SIGN_IN)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RC_SIGN_IN) {
            val task: Task<GoogleSignInAccount> = GoogleSignIn.getSignedInAccountFromIntent(data)
            handleSignInResult(task)
        }
    }

    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>) {
        try {
            val account: GoogleSignInAccount = completedTask.getResult(ApiException::class.java)
            initializeSheetsApi(account)
        } catch (e: ApiException) {
            Log.w(TAG, "signInResult:failed code=" + e.statusCode)
        }
    }

    private fun initializeSheetsApi(account: GoogleSignInAccount) {
        credential = GoogleAccountCredential.usingOAuth2(
            applicationContext, listOf(SheetsScopes.SPREADSHEETS)
        )
        credential.selectedAccount = account.account

        sheetsService = Sheets.Builder(
            NetHttpTransport(),
            GsonFactory.getDefaultInstance(),
            credential
        )
            .setApplicationName("Jupiter_Reporting")
            .build()
    }

    companion object {
        private const val TAG = "MainActivity"
        private const val RC_SIGN_IN = 1000
    }
}
